File: drivers/ts/driver.ts
// Crucible TypeScript driver — persistent replay front-end for the differential loop.
//
// Speaks drivers/common/CONTRACT.md: reads length-prefixed records on stdin,
// decodes each into the probe message via the generated corelib-ts code, and
// writes one canonical line (oracle/canonical.md) per record to stdout.
//
// build.sh bundles this + the generated message.ts + corelib-ts SOURCE into one
// CJS file via esbuild (aliasing @sofa-buffers/corelib to the corelib's
// src/index.ts), so the run needs no separate corelib build and does not depend
// on the corelib's committed dist.
//
// Like Python/Java (and unlike Rust/C++), the generated TS decode throws on
// non-COMPLETE input (SofabError), so the verdict is a plain try/catch. The
// one-shot Probe.decode path (MESSAGE_SPEC §7) throws a SofabError whose .code
// distinguishes the two non-COMPLETE outcomes: Incomplete (truncation — decode
// ended mid-field, not an error → canonical "I") vs InvalidMsg (malformed →
// "R invalid_msg"). COMPLETE returns normally (→ "A <hex>").
import { readFileSync } from "node:fs";

import { Probe, ProbeDecoder } from "./message";
import { DecodeStatus, OStream, SofabError, SofabErrorCode } from "@sofa-buffers/corelib";

// --- the streaming axes (drivers/common/CONTRACT.md, "The streaming axes") --------
//
// The replay protocol hands each record over whole and re-encodes it with one call, so
// neither streaming surface of the generated API is reachable through it. Five
// environment variables open them; unset, each is today's behaviour byte for byte.
//
// TypeScript is where crucible#132 found three bugs while building the chunked decoder
// — a visitor callback that was simply not implemented (an unimplemented optional
// callback is a NO-OP in TS, so the field silently vanished), and two more — none of
// which crashed, failed to compile, or turned the conformance suite red. That is the
// class of defect this axis exists for.
//
// This backend has ONE encode surface: `serialize(os)`. There is no `encode()` and no
// `encodeTo()`, so `SOFAB_ENCODE=new|to` must be a hard error rather than a fallback —
// reporting a mode as passing that never ran is the failure the encode gate exists to
// prevent (meta: encode_surfaces=stream).
const _num = (name: string): number => {
  const v = process.env[name];
  return v ? parseInt(v, 10) || 0 : 0;
};
const _SPLIT = _num("SOFAB_SPLIT");
const _CHUNK = _num("SOFAB_CHUNK");
const _FLUSH = _num("SOFAB_FLUSH");
const _SCRUB = !!process.env.SOFAB_CHUNK_SCRUB && process.env.SOFAB_CHUNK_SCRUB !== "0";
const _ENCODE = process.env.SOFAB_ENCODE ?? "";
if (_ENCODE && _ENCODE !== "stream") {
  process.stderr.write(
    `crucible-ts: SOFAB_ENCODE=${_ENCODE} — this backend has only 'stream' ` +
      `(no encode(), no encodeTo())\n`,
  );
  process.exit(2);
}
// Announce the configuration on stderr (never parsed). Without it a driver that
// silently ignored the variables would be indistinguishable from one that honours
// them — identical stdout either way — which is the vacuous pass the gate's opt-in
// roster guards against. This is the same guard at the driver level.
if (_SPLIT || _CHUNK || _SCRUB || _FLUSH || _ENCODE) {
  process.stderr.write(
    `crucible-ts: streaming cfg split=${_SPLIT} chunk=${_CHUNK} ` +
      `scrub=${_SCRUB ? 1 : 0} enc=${_ENCODE || "stream"} flush=${_FLUSH}\n`,
  );
}

// How the record is cut on its way into the decoder. Never an empty chunk: k<=0,
// k>=len and n>=len all mean one chunk holding the whole record. A zero-length record
// yields no chunks and is not fed at all.
function chunksOf(data: Uint8Array): Uint8Array[] {
  const len = data.length;
  if (len === 0) return [];
  if (_CHUNK > 0) {
    const out: Uint8Array[] = [];
    for (let o = 0; o < len; o += _CHUNK) out.push(data.subarray(o, Math.min(o + _CHUNK, len)));
    return out;
  }
  if (_SPLIT > 0 && _SPLIT < len) return [data.subarray(0, _SPLIT), data.subarray(_SPLIT)];
  return [data];
}

// --- materialized value dump (oracle/materialized.md), SOFAB_MATERIALIZE=1 -------
// Instead of "A <hex(re-encode)>", emit "A <dump(decoded value)>": every field and
// array element made explicit. The generated Probe carries no schema type tag
// (fp32 vs fp64, unsigned vs signed) and no struct/array shape, so the walk is
// driven by the GENERATED schema descriptor (engine/structured/schema.py, committed
// as oracle/materialized-schema.json) loaded at runtime — the same generic source
// the C driver uses. Only leaf FORMATTING stays type-specific: fp32 is a JS double
// at rest and MUST be repacked through Float32 to recover its 32-bit pattern
// (canonical.md NaN-payload caveat applies); string byte length uses UTF-8 bytes.
const _MATERIALIZE = process.env.SOFAB_MATERIALIZE === "1";

// Descriptor node shapes (oracle/materialized-schema.json). Leaves carry only a
// kind; struct carries child fields; array/wrapper carry an element type + count.
interface SchemaNode {
  id: number;
  name: string;
  kind: "u" | "s" | "fp32" | "fp64" | "string" | "blob" | "struct" | "array" | "wrapper" | "struct_wrapper";
  fields?: SchemaNode[];
  elem?: "u" | "s" | "fp32" | "fp64" | "string" | "blob";
  count?: number;
}
interface SchemaDescriptor { message: string; fields: SchemaNode[]; }

// Loaded once at startup, only in materialize mode (the default/round-trip path
// never touches the schema — it stays schema-agnostic).
const _DESC: SchemaDescriptor | null = _MATERIALIZE
  ? (JSON.parse(readFileSync(
      process.env.SOFAB_MATERIALIZE_SCHEMA ?? "oracle/materialized-schema.json",
      "utf8",
    )) as SchemaDescriptor)
  : null;

function _hex(bytes: Uint8Array): string {
  let out = "";
  for (const b of bytes) out += b.toString(16).padStart(2, "0");
  return out;
}
function _f32(x: number): string {
  const buf = new ArrayBuffer(4);
  new DataView(buf).setFloat32(0, x);
  return "f" + new DataView(buf).getUint32(0).toString(16).padStart(8, "0");
}
// CORELIB_PLAN §6.5: TS has no fp32 value type, so `x` above is already a widened
// double and the widening SETS the quiet bit — an fp32 signaling NaN can never be
// recovered from it. The generated Probe therefore captures the wire bytes
// alongside the value (`<field>Fp32Raw`, populated on decode when the value is a
// NaN); the round-trip path already re-encodes from them, which is why the
// round-trip oracle sees nothing. The materialized walk is a bit-exact consumer too
// (§6.5 "Testing"), so it must read the same raw channel rather than repack the
// double. `off` selects the element inside an fp32 *array*'s flat payload.
function _f32FromRaw(raw: Uint8Array, off: number): string {
  const bits = (raw[off] | (raw[off + 1] << 8) | (raw[off + 2] << 16) |
                (raw[off + 3] << 24)) >>> 0;   // little-endian wire order
  return "f" + bits.toString(16).padStart(8, "0");
}
function _f64(x: number): string {
  const buf = new ArrayBuffer(8);
  new DataView(buf).setFloat64(0, x);
  return "F" + new DataView(buf).getBigUint64(0).toString(16).padStart(16, "0");
}
function _t(s: string): string {
  const b = Buffer.from(s, "utf-8");
  return "t" + b.length + ":" + b.toString("hex");
}
function _b(bytes: Uint8Array): string {
  return "b" + bytes.length + ":" + _hex(bytes);
}

// The one schema-specific piece: format a single leaf value per its kind. Reused for
// both scalar leaves and array/wrapper elements (the descriptor's `elem` is a leaf
// kind). u/s are number|bigint (bigint for 64-bit) → decimal via toString().
function formatLeaf(kind: string, v: unknown, raw?: unknown, off = 0): string {
  switch (kind) {
    case "u": return "u" + (v as number | bigint).toString();
    case "s": return "s" + (v as number | bigint).toString();
    case "fp32":
      // Prefer the raw wire bytes when the generated type captured them (NaN only);
      // otherwise the widened double is lossless and repacking is exact.
      return raw instanceof Uint8Array && off + 4 <= raw.length
        ? _f32FromRaw(raw, off)
        : _f32(v as number);
    case "fp64": return _f64(v as number);
    case "string": return _t(v as string);
    case "blob": return _b(v as Uint8Array);
    default: throw new Error("unhandled leaf kind " + kind);
  }
}

// Generic recursive walk: descriptor node + the corresponding in-memory value → the
// materialized-form string. No schema shape is baked in here — structs, arrays, and
// wrappers are all discovered from the node.
function walk(node: SchemaNode, value: unknown, raw?: unknown): string {
  switch (node.kind) {
    case "struct": {
      const v = value as Record<string, unknown>;
      // Thread the sibling raw-bits field (§6.5) down with each child: the generated
      // type parks it next to the value as `<name>Fp32Raw`, for a scalar fp32 and for
      // an fp32 array alike (there it is the flat count*4 payload).
      return "{" + node.fields!.map((c) =>
        c.id + ":" + walk(c, v[c.name], v[c.name + "Fp32Raw"])).join(";") + "}";
    }
    case "array":
    case "wrapper":
      // array: numeric/fp materialized to N in memory; wrapper: index order,
      // container length is the signal. Both just map over the in-memory elements.
      return "[" + (value as unknown[]).map((el, i) =>
        formatLeaf(node.elem!, el, raw, i * 4)).join(",") + "]";
    case "struct_wrapper":
      // struct_array (WP-05): elements are generated objects — an obj walk per
      // element, container length as-is (like `wrapper`).
      return "[" + (value as Record<string, unknown>[]).map((e) =>
        "{" + node.fields!.map((c) => c.id + ":" + walk(c, e[c.name])).join(";") + "}"
      ).join(",") + "]";
    default:
      return formatLeaf(node.kind, value, raw);
  }
}

function materialize(m: Probe): string {
  const d = _DESC!;
  const v = m as unknown as Record<string, unknown>;
  return "{" + d.fields.map((c) => c.id + ":" + walk(c, v[c.name])).join(";") + "}";
}

function rejectClass(e: unknown): string {
  // Coarse in Phase 2 (reject-class comparison is soft per policy). A SofabError
  // is a decode-level failure; anything else is surfaced as "other" rather than
  // hidden.
  return e instanceof SofabError ? "invalid_msg" : "other";
}

// Re-encode through the streaming surface, which is the only one this backend has.
// SOFAB_FLUSH=n gives the OStream an n-byte buffer draining to a sink, so the message
// arrives in n-byte pieces and the encoder crosses a buffer boundary at every offset —
// the encode-side mirror of SOFAB_CHUNK=1. The bytes must not change either way.
function encodeBytes(m: Probe): Uint8Array {
  if (_FLUSH > 0) {
    const parts: Uint8Array[] = [];
    const os = new OStream(new Uint8Array(_FLUSH), 0, (c) => parts.push(Uint8Array.from(c)));
    try {
      m.serialize(os);
    } catch (e) {
      // corelib-ts's OStream.ensure(n) needs n CONTIGUOUS bytes and only flushes
      // before checking, so a caller buffer smaller than the largest single write
      // (e.g. 4 for an fp32, count*10 for a varint array) cannot encode at all. That
      // is a property of this backend, not of the message: corelib-cpp streams the
      // same value through a 1-byte buffer. Exit 3 — distinct from 2, "the backend
      // has no such surface" — so the gate can report the size as inapplicable
      // instead of either failing or, worse, silently skipping it.
      if (e instanceof SofabError && e.code === SofabErrorCode.BufferFull) {
        process.stderr.write(
          `crucible-ts: SOFAB_FLUSH=${_FLUSH} is below this backend's contiguous-write ` +
            `requirement (OStream.ensure needs n contiguous bytes): ${e.message}\n`,
        );
        process.exit(3);
      }
      throw e;
    }
    os.flush();
    const total = parts.reduce((n, p) => n + p.length, 0);
    const out = new Uint8Array(total);
    let o = 0;
    for (const p of parts) { out.set(p, o); o += p.length; }
    return out;
  }
  const os = new OStream();
  m.serialize(os);
  return os.bytes();
}

// Chunked decode via the generated ProbeDecoder (crucible#132; new in sofabgen
// cfe5250b). Used ONLY when a chunking variable is set, so the default path stays the
// one-shot Probe.decode byte for byte — which is also what makes the comparison
// meaningful: the gate then checks `decode(whole) == feed(a); feed(b); …`, two
// genuinely different code paths, rather than one path against itself.
//
// The verdict comes from `status`, never from `finish()`: finish() throws mid-field
// here and returns null in Dart, so routing the verdict through it would bake a
// backend difference into the canonical line (CONTRACT.md).
function canonicalChunked(data: Uint8Array): string {
  const d = new ProbeDecoder();
  try {
    for (const c of chunksOf(data)) {
      // Scrub mode needs a buffer the driver owns: feed it, then overwrite. A decoder
      // that borrowed from the chunk instead of copying out of it reads back 0xA5.
      const buf = _SCRUB ? Uint8Array.from(c) : c;
      d.feed(buf);
      if (_SCRUB) buf.fill(0xa5);
    }
  } catch (e) {
    if (e instanceof SofabError && e.code === SofabErrorCode.Incomplete) return "I";
    if (e instanceof SofabError && e.code === SofabErrorCode.LimitExceeded) return "L";
    return "R " + rejectClass(e);
  }
  const st = d.status;
  if (st === DecodeStatus.Invalid) return "R invalid_msg";
  if (st !== DecodeStatus.Complete) return "I";
  const m = d.message;
  if (_MATERIALIZE) return "A " + materialize(m);
  return "A " + _hex(encodeBytes(m));
}

function canonical(data: Uint8Array): string {
  // decode -> re-encode -> hex (oracle/canonical.md). The generated TS message
  // has no encode(), so serialize into an in-memory OStream and read its bytes.
  if (_SPLIT || _CHUNK || _SCRUB) return canonicalChunked(data);
  let m: Probe;
  try {
    m = Probe.decode(data);
  } catch (e) {
    // INCOMPLETE (truncation) is a distinct hard verdict, not a reject: the
    // stream ended inside a field. Detect it first (canonical.md: never collapse
    // it into A or R). The optional partial-value hex payload is not emitted —
    // the throwing one-shot path yields no partial value, and the payload axis is
    // soft in Phase 2.
    if (e instanceof SofabError && e.code === SofabErrorCode.Incomplete) {
      return "I";
    }
    if (e instanceof SofabError && e.code === SofabErrorCode.LimitExceeded) {
      // LIMIT_EXCEEDED (generator#102, limit mode only): a configured receiver-side
      // cap on a schema-unbounded field. A policy rejection distinct from INVALID —
      // its own verdict `L`, not `R`.
      return "L";
    }
    return "R " + rejectClass(e);
  }
  // COMPLETE decode. In materialize mode, dump the decoded value (oracle/
  // materialized.md) instead of the re-encoded wire hex; the default path is
  // unchanged.
  if (_MATERIALIZE) {
    return "A " + materialize(m);
  }
  return "A " + _hex(encodeBytes(m));
}

function main(): void {
  const input = readFileSync(0); // whole stdin (comparator writes all frames, then EOF)
  const lines: string[] = [];
  let off = 0;
  while (off + 4 <= input.length) {
    const n = input.readUInt32LE(off);
    off += 4;
    if (off + n > input.length) {
      process.stderr.write("crucible-ts: short payload\n");
      process.exit(1);
    }
    lines.push(canonical(input.subarray(off, off + n)));
    off += n;
  }
  process.stdout.write(lines.length ? lines.join("\n") + "\n" : "");
}

main();

File: drivers/dart/driver.dart
// Crucible Dart driver — persistent replay front-end for the differential loop.
//
// Speaks drivers/common/CONTRACT.md: reads length-prefixed records from stdin
// (a stream of <u32 little-endian len><N payload bytes>), decodes each into the
// probe message via the generated corelib-dart code, and writes one canonical
// line per record (oracle/canonical.md) to stdout, in input order.
//
// Decode is status-returning (MESSAGE_SPEC §7): Probe.tryDecode fills the message
// and returns the terminal DecodeStatus, with schema-bound violations (over-count,
// over-index, over-maxlen) folded into `invalid` via the generated sticky flag.
// The mapping to the canonical verdict is 1:1 — complete→A, incomplete→I,
// invalid→R, limitExceeded→L — the same shape as the Go driver.
//
// The Dart coverage engine (dart:ffi shared lib + a C libFuzzer harness, like
// Zig) is a separate front-end; this file is the replay path only.
import 'dart:io';
import 'dart:typed_data';

import 'package:sofabuffers/sofabuffers.dart' as sofab;
import 'message.dart';
import 'materialize_gen.dart';

// SOFAB_MATERIALIZE=1 selects the materialized-value dump (oracle/materialized.md)
// on a COMPLETE decode instead of the round-trip hex; unset keeps the default
// round-trip path. Only the `A` payload changes.
final bool _materialize = Platform.environment['SOFAB_MATERIALIZE'] == '1';

const _hexDigits = '0123456789abcdef';

String _hex(Uint8List b) {
  final sb = StringBuffer();
  for (final x in b) {
    sb.write(_hexDigits[(x >> 4) & 0xf]);
    sb.write(_hexDigits[x & 0xf]);
  }
  return sb.toString();
}

// ---- the streaming axes (drivers/common/CONTRACT.md) -------------------------
//
// The replay protocol hands each record over whole and re-encodes it with one call,
// so neither streaming surface of the generated API is reachable through it. Unset,
// every variable below is today's behaviour byte for byte.
//
// Dart is the backend the contract's finish() rule was written for: ProbeDecoder's
// finish() returns NULL where the others throw, so routing the verdict through it
// would put a backend difference into the canonical line. It exposes `status`, so
// the verdict comes from there.
int _envInt(String name) => int.tryParse(Platform.environment[name] ?? '') ?? 0;

final int _split = _envInt('SOFAB_SPLIT');
final int _chunk = _envInt('SOFAB_CHUNK');
final int _flush = _envInt('SOFAB_FLUSH');
final bool _scrub = (Platform.environment['SOFAB_CHUNK_SCRUB'] ?? '').isNotEmpty &&
    Platform.environment['SOFAB_CHUNK_SCRUB'] != '0';
final String _encode = (Platform.environment['SOFAB_ENCODE'] ?? '').isEmpty
    ? 'new'
    : Platform.environment['SOFAB_ENCODE']!;
final bool _chunking = _split != 0 || _chunk != 0 || _scrub;

void _checkCfg() {
  if (_encode != 'new' && _encode != 'to' && _encode != 'stream') {
    stderr.writeln('crucible-dart: unknown SOFAB_ENCODE=$_encode '
        '(this backend has new, to, stream)');
    exit(2);
  }
  // Announce on stderr (never parsed). A driver that silently ignored these would
  // be indistinguishable from one that honours them — stdout is identical either
  // way — so this makes "it really re-feeds" checkable rather than asserted.
  if (_split != 0 || _chunk != 0 || _scrub || _flush != 0 || _encode != 'new') {
    stderr.writeln('crucible-dart: streaming cfg split=$_split chunk=$_chunk '
        'scrub=${_scrub ? 1 : 0} enc=$_encode flush=$_flush');
  }
}

// How the record is cut on its way in. Never an empty chunk; a 0-byte record
// yields none at all.
List<Uint8List> _chunksOf(Uint8List data) {
  final len = data.length;
  if (len == 0) return const [];
  if (_chunk > 0) {
    final out = <Uint8List>[];
    for (var o = 0; o < len; o += _chunk) {
      out.add(Uint8List.fromList(data.sublist(o, o + _chunk > len ? len : o + _chunk)));
    }
    return out;
  }
  if (_split > 0 && _split < len) {
    return [
      Uint8List.fromList(data.sublist(0, _split)),
      Uint8List.fromList(data.sublist(_split)),
    ];
  }
  return [Uint8List.fromList(data)];
}

// Re-encode through the surface SOFAB_ENCODE selects. All three must emit identical
// bytes, and SOFAB_FLUSH must not change them either: it gives the Encoder an
// n-byte buffer draining to a sink, so the encoder crosses a buffer boundary at
// every offset.
Uint8List _encodeVia(Probe m) {
  if (_encode == 'new') return m.encode();
  final builder = BytesBuilder(copy: true);
  // corelib-dart#62 dropped `bufferSize:`: per CORELIB_PLAN §5.1 the corelib
  // allocates no output buffer, so the caller supplies one. A buffer below
  // `minOutputBuffer` is refused at the handover with `invalidArgument` — report
  // that as exit 3 ("cannot operate at this configuration") rather than letting
  // the exception escape, so the oracle sees the size the port declined.
  final cap = _flush > 0 ? _flush : 4096;
  final sofab.Encoder enc;
  try {
    enc = sofab.Encoder(builder.add, buffer: Uint8List(cap));
  } on sofab.SofabException catch (e) {
    stderr.writeln('crucible-dart: Encoder refused a $cap-byte buffer '
        '(MIN_OUTPUT_BUFFER=${sofab.minOutputBuffer}): $e');
    exit(3);
  }
  if (_encode == 'to') {
    m.encodeTo(enc);
  } else {
    m.serialize(enc);
  }
  enc.flush();
  return builder.toBytes();
}

// canonical produces the one canonical line for a single candidate input
// (oracle/canonical.md: decode -> re-encode -> hex on COMPLETE).
String canonical(Uint8List data) {
  final out = Probe();
  final sofab.DecodeStatus st;
  try {
    if (_chunking) {
      // Chunked decode via the generated ProbeDecoder, taken ONLY when a chunking
      // variable is set — the default stays the one-shot tryDecode byte for byte,
      // which is what makes the gate meaningful: it then compares two genuinely
      // different code paths. Verdict from `status`, never from finish().
      final d = Probe.decoder(out);
      for (final c in _chunksOf(data)) {
        d.feed(c);
        // Scrub: the chunk is a copy the driver owns, so overwriting it after feed
        // exposes a decoder that borrowed instead of copying.
        if (_scrub) c.fillRange(0, c.length, 0xA5);
      }
      st = d.status;
    } else {
      st = Probe.tryDecode(data, out);
    }
  } catch (_) {
    // The generated tryDecode is not expected to throw on any input; if it does
    // that is itself worth surfacing, mapped to the coarse `other` reject class.
    return 'R other';
  }

  // INCOMPLETE (MESSAGE_SPEC §7): bytes end inside a field/varint or an open
  // sequence — the third verdict, neither accept nor reject. corelib-dart returns
  // no partial value here, so emit the bare `I`.
  if (st == sofab.DecodeStatus.incomplete) return 'I';
  // LIMIT_EXCEEDED (generator#102, limit mode only): a configured receiver-side
  // cap on a schema-unbounded field. A policy rejection distinct from INVALID.
  if (st == sofab.DecodeStatus.limitExceeded) return 'L';
  // INVALID: malformed regardless of what follows. Reject class is coarse
  // (soft axis, see oracle/policy.yaml) — the status carries no finer code.
  if (st == sofab.DecodeStatus.invalid) return 'R invalid_msg';

  // COMPLETE. In materialize mode, dump the decoded value (oracle/materialized.md);
  // otherwise re-encode with the corelib's own sparse-canonical encoder and emit the
  // lowercase hex (schema-agnostic; folds in the round-trip oracle).
  if (_materialize) return 'A ${materialize(out)}';
  try {
    return 'A ${_hex(_encodeVia(out))}';
  } catch (_) {
    return 'R other';
  }
}

Future<void> main() async {
  _checkCfg();
  // Read the whole framed stream (the comparator writes it all, then reads our
  // stdout after we exit — see oracle/comparator.py run_driver). The corpus fits
  // in memory; this mirrors the TypeScript driver.
  final bb = BytesBuilder(copy: false);
  await for (final chunk in stdin) {
    bb.add(chunk);
  }
  final data = bb.toBytes();
  final bd = ByteData.sublistView(data);

  final sb = StringBuffer();
  var off = 0;
  while (off + 4 <= data.length) {
    final n = bd.getUint32(off, Endian.little);
    off += 4;
    if (off + n > data.length) break; // incomplete trailing frame — stop cleanly
    final rec = Uint8List.sublistView(data, off, off + n);
    off += n;
    sb.writeln(canonical(rec));
  }
  stdout.write(sb.toString());
  await stdout.flush();
}
